pkg/runtime/wfengine/backend.go

```diff
@@ -0,0 +1,267 @@
+	wfLogger.Warnf("%s: aborting activity execution (#%d)", wi.InstanceID, wi.NewEvent.EventId)
+
+	// Sending false signals the waiting activity actor to abort the activity execution.
+	if channel, ok := wi.Properties[CallbackChannelProperty]; ok {
+		channel.(chan bool) <- false
+	}
+	return nil
+}
+
+// AbandonOrchestrationWorkItem implements backend.Backend. It gets called by durabletask-go when there is
+// an unexpected failure in the workflow orchestration execution pipeline.
+func (*actorBackend) AbandonOrchestrationWorkItem(ctx context.Context, wi *backend.OrchestrationWorkItem) error {
+	wfLogger.Warnf("%s: aborting workflow execution", wi.InstanceID)
+
+	// Sending false signals the waiting workflow actor to abort the workflow execution.
+	if channel, ok := wi.Properties[CallbackChannelProperty]; ok {
+		channel.(chan bool) <- false
+	}
+	return nil
+}
+
+// AddNewOrchestrationEvent implements backend.Backend and sends the event e to the workflow actor identified by id.
+func (be *actorBackend) AddNewOrchestrationEvent(ctx context.Context, id api.InstanceID, e *backend.HistoryEvent) error {
+	data, err := backend.MarshalHistoryEvent(e)
+	if err != nil {
+		return err
+	}
+
+	// Send the event to the corresponding workflow actor, which will store it in its event inbox.
+	req := invokev1.
+		NewInvokeMethodRequest(AddWorkflowEventMethod).
+		WithActor(WorkflowActorType, string(id)).
+		WithRawDataBytes(data).
+		WithContentType(invokev1.OctetStreamContentType)
+	defer req.Close()
+
+	resp, err := be.actors.Call(ctx, req)
+	if err != nil {
+		return err
+	}
+	defer resp.Close()
+	return nil
+}
+
+// CompleteActivityWorkItem implements backend.Backend
+func (*actorBackend) CompleteActivityWorkItem(ctx context.Context, wi *backend.ActivityWorkItem) error {
+	// Sending true signals the waiting activity actor to complete the execution normally.
+	wi.Properties[CallbackChannelProperty].(chan bool) <- true
+	return nil
+}
+
+// CompleteOrchestrationWorkItem implements backend.Backend
+func (*actorBackend) CompleteOrchestrationWorkItem(ctx context.Context, wi *backend.OrchestrationWorkItem) error {
+	// Sending true signals the waiting workflow actor to complete the execution normally.
+	wi.Properties[CallbackChannelProperty].(chan bool) <- true
+	return nil
+}
+
+// CreateTaskHub implements backend.Backend
+func (*actorBackend) CreateTaskHub(context.Context) error {
+	return nil
+}
+
+// DeleteTaskHub implements backend.Backend
+func (*actorBackend) DeleteTaskHub(context.Context) error {
+	return errors.New("not supported")
+}
+
+// GetActivityWorkItem implements backend.Backend
+func (be *actorBackend) GetActivityWorkItem(ctx context.Context) (*backend.ActivityWorkItem, error) {
+	// Wait for the workflow actor to signal us with some work to do
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case wi := <-be.activityWorkItemChan:
+		return wi, nil
```

Comment:
what if the channel got closed, is it ok to process a zero-value activity work item?

Reply:
It's not, but I never close this channel. Is there a particular best practice that I should be following regarding managing channel lifetimes in this case?

Reply:
Theoretically the channel lifecycle is tied to the receiver, the receiver should always be responsible for closing the channel, and you can't do any assumptions like "this channel should never be closed" as you can't express it compile time (e.g using a channel type that can't be closed), so I'd rather handling closed channels just in case

Reply:
Personally, I like to be cautious about doing "just in case" development because it can confuse the actual intent of the code. For example, if I'm new to the code base and see handling for a channel being closed, then I will naturally assume that the channel is _expected_ to be closed at some point. However, that's not actually true, which creates confusion for me as the reader. For that reason, I'd prefer we add handling for closed channels at the point we decide to start closing channels.

Reply:
I think it's ok to keep it that way as long as you have a unit test that can catch the problem if the channel is closed erroneously. Otherwise you didn't add code here because of "just in case" and also failed to prevent an erroneously closed channel.

Reply:
I don't see much value in writing a test for a condition that's never expected to happen. It seems like a lot more work and just as confusing. I will go ahead and just add some logic to handle the channel being unexpectedly closed.

Reply:
Ok, you can leave it the way it is if it takes all this time to do the test, even the test is preferable in this case.